src/epoch_credits/mod.rs

```diff
@@ -1,35 +1,176 @@
-
-    // Convert epoch number to slot
-    let start_slot = epoch * info.slots_in_epoch;
-
-    // We cannot use an excessively large range if the epoch just started. There is a chance that
-    // the end slot has not been reached and strange behaviour will occur.
-    // If this is the current epoch and less than `SLOT_OFFSET` slots have elapsed, then do not define an
-    // end_slot for use in the RPC call.
-    let end_slot = if info.epoch == epoch && info.slot_index < SLOT_OFFSET {
-        None
-    } else {
-        Some(start_slot + SLOT_OFFSET)
-    };
-
-    // First block only
-    let block = client
-        .get_blocks(start_slot, end_slot)?
-        .get(0)
-        .cloned()
-        .ok_or_else(|| anyhow!("no blocks found"))?;
-
-    Ok(client.get_block(block)?.rewards)
+struct StakingApy {
+    voter: String,
+    percent: f64,
+}
+
+struct ValidatorReward {
+    voter: String,
+    lamports: u64,
+}
+
+/// The monitor of rewards paid to validators and delegators.
+pub struct RewardsMonitor<'a> {
+    /// Shared Solana RPC client.
+    client: &'a RpcClient,
+    /// Prometheus staking APY gauge.
+    staking_apys: &'a GaugeVec,
+    /// Prometheus cumulative validator rewards gauge.
+    validator_rewards: &'a GaugeVec,
+    /// The last epoch in which rewards were recorded. 0 if uninitialised.
+    ///
+    /// FIXME: cached epoch credit history.
+    last_rewards_epoch: Epoch,
+}
+
+impl<'a> RewardsMonitor<'a> {
+    /// Initialises a new rewards monitor.
+    pub fn new(
+        client: &'a RpcClient,
+        staking_apys: &'a GaugeVec,
+        validator_rewards: &'a GaugeVec,
+    ) -> Self {
+        Self {
+            client,
+            staking_apys,
+            validator_rewards,
+            last_rewards_epoch: 0,
+        }
+    }
+
+    /// Exports reward metrics once an epoch.
+    pub fn export_rewards(&mut self, epoch_info: &EpochInfo) -> anyhow::Result<()> {
+        let epoch = epoch_info.epoch;
+        // FIXME: check seen epochs in the cache.
+        if epoch <= self.last_rewards_epoch {
+            return Ok(());
+        }
+        if let Some(rewards) = self.get_rewards_for_epoch(epoch, Some(epoch_info.clone()))? {
+            // FIXME: replace 3.0 with the previous epoch duration when using only one epoch, and
+            // with the average of all used epochs if using several.
+            let epochs_in_year = 365.0 / 3.0;
+            let (staking_apys, validator_rewards) =
+                self.process_rewards(rewards, epochs_in_year)?;
+
+            for s in staking_apys {
+                self.staking_apys
+                    .get_metric_with_label_values(&[&s.voter])
+                    .map(|c| c.set(s.percent))?;
+            }
+
+            for v in validator_rewards {
+                self.validator_rewards
+                    .get_metric_with_label_values(&[&v.voter])
+                    .map(|c| c.set(v.lamports as f64 / LAMPORTS_PER_SOL as f64))?;
```

Comment:
Beware floating-point representation - this division may end up giving us an incorrect value. We should do all calculation in lamports and then perform the division on the frontend.

Reply:
The rounding error would be negligible though. We can use `IntGaugeVec`, although that still has theoretical issues due to the conversion of lamports of type `u64` to `i64`, into the upper half of it. Not that we should be worried about this.